bsp/src/mill/bsp/MillBspLogger.scala

```diff
@@ -1,57 +1,86 @@
+import java.nio.charset.Charset
+import mill.bsp.MillBspLogger.tickerPattern
+import mill.util.ColorLogger
 
-  override def ticker(s: String): Unit = {
-    try {
-      val progressString = s.split(" ")(0)
-      val progress = progressString.substring(1, progressString.length - 1).split("/")
-      val params = new TaskProgressParams(new TaskId(taskId.toString))
-      params.setEventTime(System.currentTimeMillis())
-      params.setMessage(s)
-      params.setUnit(s.split(" ")(1))
-      params.setProgress(progress(0).toLong)
-      params.setTotal(progress(1).toLong)
-      client.onBuildTaskProgress(params)
-      super.ticker(s)
-    } catch {
-      case e: Exception =>
-    }
+object MillBspLogger {
+  /**
+   * Creates a BSP-specialized logger class which sends `task-progress`
+   * notifications ( upon the invocation of the `ticker` method ) and
+   * `show-message` notifications ( for each error or information
+   * being logged ).
+   *
+   * @param client the client to send notifications to, also the
+   *               client that initiated a request which triggered
+   *               a mill task evaluation
+   * @param taskId unique ID of the task being evaluated
+   * @param inputStream The input stream to read input from
+   */
+  def createBspLogger(client: BuildClient, taskId: Int, inputStream: InputStream): ColorLogger = {
+    val bspErrorStream = new MillBspClientPrintStream(client, MessageType.INFORMATION, new ByteArrayOutputStream())
+    val bspOutputStream = new MillBspClientPrintStream(client, MessageType.INFORMATION, new ByteArrayOutputStream())
+    new MillBspLogger(client, taskId, inputStream, bspOutputStream, bspErrorStream)
+  }
+
+  private val tickerPattern = """\[(?<progress>\d+)/(?<total>\d+)] (?<unit>.\S+)\s+$""".r
+}
+
+private class MillBspLogger(
+                      client: BuildClient,
+                      taskId: Int,
+                      inputStream: InputStream,
+                      outStream: PrintStream,
+                      errStream: PrintStream) extends ColorLogger {
+
+  override def colored: Boolean = false
+
+  override def colors: Colors = Colors.BlackWhite
+
+  override def errorStream: PrintStream = errStream
+
+  override def outputStream: PrintStream = outStream
+
+  override def inStream: InputStream = inputStream
+
+  override def info(s: String): Unit = {
+    client.onBuildShowMessage(new ShowMessageParams(MessageType.INFORMATION, s))
   }
 
   override def error(s: String): Unit = {
-    super.error(s)
     client.onBuildShowMessage(new ShowMessageParams(MessageType.ERROR, s))
   }
 
-  override def info(s: String): Unit = {
-    super.info(s)
-    client.onBuildShowMessage(new ShowMessageParams(MessageType.INFORMATION, s))
+  override def ticker(s: String): Unit = {
+    try {
+      client.onBuildShowMessage(new ShowMessageParams(MessageType.INFORMATION, s))
+      tickerPattern.findFirstMatchIn(s).foreach { ticker =>
```

Comment:
This ticker implementation is broken in a slightly different way from how it was broken previously. See https://github.com/lihaoyi/mill/issues/1034. Fixing that will likely require changing this again.